File: backend/src/main/java/com/smartcampost/backend/service/DeliveryService.java
package com.smartcampost.backend.service;

import com.smartcampost.backend.dto.delivery.*;

import java.util.UUID;

/**
 * Service for managing the complete delivery workflow.
 * Orchestrates OUT_FOR_DELIVERY, OTP verification, and proof capture.
 */
public interface DeliveryService {

    /**
     * Start delivery process - marks parcel as OUT_FOR_DELIVERY.
     * Courier scans parcel QR at destination agency before going out.
     */
    StartDeliveryResponse startDelivery(StartDeliveryRequest request);

    /**
     * Send OTP to recipient before delivery.
     * Triggered when courier is near delivery location.
     */
    void sendDeliveryOtp(UUID parcelId);

    /**
     * Send/re-send OTP to recipient and record an OTP_SENT ScanEvent (GPS required).
     */
    void sendDeliveryOtp(UUID parcelId, Double latitude, Double longitude, String notes);

    /**
     * Complete delivery with full verification.
     * Verifies OTP, captures proof, and marks parcel as DELIVERED.
     */
    CompleteDeliveryResponse completeDelivery(CompleteDeliveryRequest request);

    /**
     * Get delivery status and details for a parcel.
     */
    DeliveryStatusResponse getDeliveryStatus(UUID parcelId);

    /**
     * Mark delivery as failed (recipient not available, refused, etc.)
     */
    DeliveryStatusResponse markDeliveryFailed(UUID parcelId, String reason, Double latitude, Double longitude, String notes);

    /**
     * Reschedule delivery for a later date.
     */
    DeliveryStatusResponse rescheduleDelivery(UUID parcelId, RescheduleDeliveryRequest request);

    /**
     * Agency pickup flow: QR + OTP verification then mark as PICKED_UP_AT_AGENCY.
     */
    PickupAtAgencyResponse pickupAtAgency(PickupAtAgencyRequest request);

    /**
     * Return to sender workflow.
     */
    DeliveryStatusResponse returnToSender(UUID parcelId, ReturnToSenderRequest request);
}

File: smartcampost_mobile/lib/services/delivery_service.dart
import 'package:smartcampost_mobile/core/api_client.dart';

class DeliveryService {
  final ApiClient _api = ApiClient();

  Future<void> startDelivery(Map<String, dynamic> data) async {
    await _api.post('/delivery/start', data: data);
  }

  Future<void> completeDelivery(Map<String, dynamic> data) async {
    await _api.post('/delivery/complete', data: data);
  }

  Future<void> sendDeliveryOtp({
    required String parcelId,
    required String phoneNumber,
    double? latitude,
    double? longitude,
  }) async {
    await _api.post(
      '/delivery/otp/send',
      data: {
        'parcelId': parcelId,
        'phoneNumber': phoneNumber,
        'latitude': ?latitude,
        'longitude': ?longitude,
      },
    );
  }

  Future<void> verifyDeliveryOtp({
    required String parcelId,
    required String otpCode,
    double? latitude,
    double? longitude,
  }) async {
    await _api.post(
      '/delivery/otp/verify',
      data: {
        'parcelId': parcelId,
        'otpCode': otpCode,
        'latitude': ?latitude,
        'longitude': ?longitude,
      },
    );
  }

  Future<Map<String, dynamic>> getDeliveryStatus(String parcelId) async {
    return _api.get<Map<String, dynamic>>('/delivery/$parcelId/status');
  }

  Future<void> markDeliveryFailed(
    String parcelId, {
    required String reason,
    required double latitude,
    required double longitude,
    String? notes,
  }) async {
    await _api.post(
      '/delivery/$parcelId/failed',
      queryParameters: {
        'reason': reason,
        'latitude': latitude.toString(),
        'longitude': longitude.toString(),
        'notes': ?notes,
      },
    );
  }
}
